kubernetes-discovery-client/src/main/java/io/micronaut/kubernetes/discovery/KubernetesDiscoveryClient.java

```diff
@@ -94,47 +94,32 @@ public KubernetesDiscoveryClient(KubernetesClient client,
-            String labelSelector = KubernetesUtils.computeLabelSelector(discoveryConfiguration.getLabels());
-            AtomicReference<Metadata> metadata = new AtomicReference<>();
-            Predicate<KubernetesObject> includesFilter = KubernetesUtils.getIncludesFilter(discoveryConfiguration.getIncludes());
-            Predicate<KubernetesObject> excludesFilter = KubernetesUtils.getExcludesFilter(discoveryConfiguration.getExcludes());
-
-            if (LOG.isTraceEnabled()) {
-                LOG.trace(String.format("Fetching for service %s endpoints in namespace %s", serviceName, serviceNamespace));
+            KubernetesServiceConfiguration serviceConfiguration = serviceConfigurations.computeIfAbsent(
+                    serviceId, key -> new KubernetesServiceConfiguration(key, null, null, false));
+
+            if (!serviceConfiguration.getNamespace().isPresent()) {
+                serviceConfiguration.setNamespace(configuration.getNamespace());
             }
 
-            return Flowable.fromPublisher(client.listEndpoints(serviceNamespace, labelSelector))
-                    .doOnError(throwable -> LOG.error("Error while trying to list Kubernetes Endpoints in the namespace [" + serviceNamespace + "]", throwable))
-                    .flatMapIterable(EndpointsList::getItems)
-                    .filter(endpoints -> endpoints.getMetadata().getName().equals(serviceName))
-                    .filter(includesFilter)
-                    .filter(excludesFilter)
-                    .doOnNext(endpoints -> metadata.set(endpoints.getMetadata()))
-                    .flatMapIterable(Endpoints::getSubsets)
-                    .map(subset -> subset
-                            .getPorts()
-                            .stream()
-                            .flatMap(port -> subset.getAddresses().stream().map(address -> buildServiceInstance(serviceId, port, address, metadata.get())))
-                            .collect(Collectors.toList()))
-                            .onErrorReturn(throwable -> new ArrayList<>());
-        }
-    }
+            if (!serviceConfiguration.getName().isPresent()) {
+                serviceConfiguration.setName(serviceId);
+            }
 
-    private ServiceInstance buildServiceInstance(String serviceId, Port port, Address address, Metadata metadata) {
-        boolean isSecure = port.isSecure() || metadata.isSecure();
-        String scheme = isSecure ? "https://" : "http://";
-        URI uri = URI.create(scheme + address.getIp().getHostAddress() + ":" + port.getPort());
-        if (LOG.isTraceEnabled()) {
-            LOG.trace("Building ServiceInstance for serviceId [{}] and URI [{}]", serviceId, uri.toString());
+            if (!serviceConfiguration.getMode().isPresent()) {
+                serviceConfiguration.setMode(configuration.getDiscovery().getMode());
+            }
+
+            String mode = serviceConfiguration.getMode().get();
```

Comment:
```suggestion
            String mode = serviceConfiguration.getMode().orElse(configuration.getDiscovery().getMode());
```

Reply:
this is what is basically in the if block above.

Reply:
I know, but is less idiomatic no? It isn't a big deal anyway, just a suggestion on what I think is a better use of optionals